types/react-redux/index.d.ts

```diff
@@ -52,6 +53,20 @@ interface AdvancedComponentDecorator<TProps, TOwnProps> {
     (component: ComponentType<TProps>): ComponentClass<TOwnProps>;
 }
 
+/**
+ * a property P will be present if:
+ * - it is present in DecorationTargetProps
+ * - it is present in InjectedProps[P] and can satisfy DecorationTargetProps[P]
+ *   or it is not present in InjectedProps[P]
+ */
+type Matching<InjectedProps, DecorationTargetProps> = {
+	[P in keyof DecorationTargetProps]: P extends keyof InjectedProps
```

Comment:
This will accept the following test case that fails on master:
```patch
index 67c21bf3d2..ff06cf338e 100644
--- a/types/react-redux/react-redux-tests.tsx
+++ b/types/react-redux/react-redux-tests.tsx
@@ -1154,6 +1154,10 @@ namespace TestFailsMoreSpecificInjectedProps {
   // Since it is possible the injected props could fail to satisfy the decoration props,
   // the following line should fail to compile.
   connect(mapStateToProps, mapDispatchToProps)(Component) // $ExpectError
+
+  const FunctionalComponent = (props: MoreSpecificDecorationProps) => null;
+  // found error on master but not in #28189
+  connect(mapStateToProps, mapDispatchToProps)(FunctionalComponent); // $ExpectError
 }
```

I narrowed the issue down to [this playground](https://bit.ly/2NhVNs0). Typescript accepts assigning a function that takes `never` to a function that takes `T`.

Reply:
Thanks @eps1lon. I've adjusted the definition to also pass that test case.

Reply:
@eps1lon: do you approve the change? We all hope it can be merged. :)